nucypher/characters/lawful.py

```diff
@@ -1279,6 +1284,20 @@ def stop(self, halt_reactor: bool = False) -> None:
         if halt_reactor:
             reactor.stop()
 
+    def _finalize(self):
+        """
+        Cleans up Ursula from objects that may eat up system resources.
+        Useful for testing purposes, where many Ursulas are created and destroyed,
+        and references to them may persist for too long.
+        This method is not needed if all references to the Ursula are released.
+
+        **Warning:** invalidates the Ursula.
+        """
+
+        # `rest_server` holds references to the datastore (directly and via `rest_app`).
+        # An open datastore hogs up file descriptors.
+        self.rest_server = INVALIDATED
+
```

Comment:
:zap: Excellent - I might need a bit of clarification on the character's lifecycle at this point considering, `start`, `stop`, `enchant`, `disenchant`, and `finalize`.

Reply:
I would love that myself. There's even a note in `gwme` that for some reason `Ursula` has `stop` and all the others have `disenchant`. Ideally, I think there should be a non-invalidating `stop` (which could be followed by `start`), and an invalidating `finalize` (used primarily in tests to speed up the cleanup), the same for all characters. Or whichever other names as long as they're consistent.

Reply:
^ @jMyles

Reply:
Opened: https://github.com/nucypher/nucypher/issues/2293